src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+
+            _consumed++;
+            _lineBytePosition++;
+            _tokenType = JsonTokenType.PropertyName;
+            return true;
+        }
+
+        private bool ConsumeString()
+        {
+            Debug.Assert(_buffer.Length >= _consumed + 1);
+            Debug.Assert(_buffer[_consumed] == JsonConstants.Quote);
+
+            //Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localCopy = _buffer;
+
+            int idx = localCopy.Slice(_consumed + 1).IndexOf(JsonConstants.Quote);
+            if (idx < 0)
+            {
+                if (IsLastSpan)
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.EndOfStringNotFound);
+                else
+                    return false;
+            }
+
+            if (localCopy[idx + _consumed] != JsonConstants.ReverseSolidus)
+            {
+                localCopy = localCopy.Slice(_consumed + 1, idx);
+
+                if (localCopy.IndexOfAnyControlOrEscape() != -1)
+                {
+                    _lineBytePosition++;
+                    ValidateEscapingAndHex(localCopy);
+                    goto Done;
+                }
+
+                _lineBytePosition += idx + 1;
+
+            Done:
+                _lineBytePosition++;
+                ValueSpan = localCopy;
+                _tokenType = JsonTokenType.String;
+                _consumed += idx + 2;
+                return true;
+            }
+            else
+            {
+                return ConsumeStringWithNestedQuotes();
+            }
+        }
+
+        private bool ConsumeStringWithNestedQuotes()
+        {
+            //TODO: Optimize looking for nested quotes
+            //TODO: Avoid redoing first IndexOf search
+            int i = _consumed + 1;
+
+            Debug.Assert(_buffer.Length >= i);
+
+            while (true)
+            {
+                int counter = 0;
+                int foundIdx = _buffer.Slice(i).IndexOf(JsonConstants.Quote);
+                if (foundIdx == -1)
+                {
+                    if (IsLastSpan)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.EndOfStringNotFound);
+                    else
+                        return false;
+                }
+                if (foundIdx == 0)
+                    break;
+                for (int j = i + foundIdx - 1; j >= i; j--)
+                {
+                    if (_buffer[j] != JsonConstants.ReverseSolidus)
+                    {
+                        if ((counter & 1) == 0)
+                        {
+                            i += foundIdx;
+                            goto FoundEndOfString;
+                        }
+                        break;
+                    }
+                    else
+                        counter++;
+                }
+                i += foundIdx + 1;
+            }
```

Comment:
I feel that this while could really use some comments.

I figured out that the `foundIdx == 0` is that if the seek ahead produces a quote as the next character that clearly it wasn't escaped.

I figured out that the `(counter & 1)` test is about even/odd for the number of backslashes (`\\"` and `\\\\"` are unescaped, `\\\"` and `\"` are escaped).

But it took a lot of mental gymnastics to figure out what this method was doing, so it seems (to me) like it would be hard for someone else to edit it if there's a suspected bug.

Reply:
Agreed. I was planning to re-write this in the near future to be more cleaner which is why I didn't comment this. But will address the feedback. Hopefully the set of tests we have so far help a bit.

Reply:
The new implementation doesn't suffer from this complexity (although it is also slower, hopefully that case is uncommon enough).